components/zcash_address/src/kind/unified.rs

```diff
@@ -0,0 +1,141 @@
+/// derived [`PartialOrd`] will sort receivers correctly. From its documentation:
+///
+/// > When derived on enums, variants are ordered by their top-to-bottom discriminant
+/// > order.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub(crate) enum Receiver {
+    Orchard([u8; 43]),
+    Sapling(kind::sapling::Data),
+    P2pkh(kind::p2pkh::Data),
+    P2sh(kind::p2sh::Data),
+    Unknown { typecode: u8, data: Vec<u8> },
+}
+
+impl TryFrom<(u8, &[u8])> for Receiver {
+    type Error = ParseError;
+
+    fn try_from((typecode, addr): (u8, &[u8])) -> Result<Self, Self::Error> {
+        match typecode {
+            0x00 => addr.try_into().map(Receiver::P2pkh),
+            0x01 => addr.try_into().map(Receiver::P2sh),
+            0x02 => addr.try_into().map(Receiver::Sapling),
+            0x03 => addr.try_into().map(Receiver::Orchard),
+            _ => Ok(Receiver::Unknown {
+                typecode,
+                data: addr.to_vec(),
+            }),
+        }
+        .map_err(|_| ParseError::InvalidEncoding)
+    }
+}
+
+impl Receiver {
+    fn typecode(&self) -> u8 {
+        match self {
+            Receiver::P2pkh(_) => 0x00,
+            Receiver::P2sh(_) => 0x01,
+            Receiver::Sapling(_) => 0x02,
+            Receiver::Orchard(_) => 0x03,
+            Receiver::Unknown { typecode, .. } => *typecode,
+        }
+    }
+
+    fn addr(&self) -> &[u8] {
+        match self {
+            Receiver::P2pkh(data) => data,
+            Receiver::P2sh(data) => data,
+            Receiver::Sapling(data) => data,
+            Receiver::Orchard(data) => data,
+            Receiver::Unknown { data, .. } => data,
+        }
+    }
+}
+
+/// A Unified Address.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Address(pub(crate) Vec<Receiver>);
+
+impl TryFrom<&[u8]> for Address {
+    type Error = ParseError;
+
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let encoded = f4jumble::f4jumble_inv(buf).ok_or(ParseError::InvalidEncoding)?;
+
+        // Validate and strip trailing zero bytes.
+        let encoded = match encoded.split_at(encoded.len() - PADDING_LEN) {
+            (encoded, tail) if tail == &[0; PADDING_LEN][..] => Ok(encoded),
+            _ => Err(ParseError::InvalidEncoding),
+        }?;
+
+        iter::repeat(())
+            .scan(encoded, |encoded, _| match encoded {
+                // Base case: we've parsed the full encoding.
+                [] => None,
+                // The raw encoding of a Unified Address is a concatenation of:
+                // - typecode: byte
+                // - length: byte
+                // - addr: byte[length]
+                [typecode, length, data @ ..] if data.len() >= *length as usize => {
+                    let (addr, rest) = data.split_at(*length as usize);
+                    *encoded = rest;
+                    Some((*typecode, addr).try_into())
```

Comment:
This `try_into` targets the `Receiver` type? This is one of those locations where type inference makes the code a bit less clear to read.

Reply:
Yes, this targets `Receiver` because the iterator is being collected into a `Vec<Receiver>` (being the type that `Address` wraps).

Reply:
```suggestion
                    let receiver: Result<Receiver, _> = (*typecode, addr).try_into();
                    Some(receiver) 
```